File: roms/pac_and_pal.py
#
#	Pac & Pal
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg1 = z.read('pap1-3b.1d') + z.read('pap1-2b.1c') + z.read('pap3-1.1b')
    prg2 = z.read('pap1-4.1k')
    bg = z.read('pap1-6.3c')
    obj = z.read('pap1-5.3f')
    rgb = z.read('pap1-6.4c')
    bgcolor = z.read('pap1-5.4e')
    objcolor = z.read('pap1-4.3l')
    snd = z.read('pap1-3.3m')

rom = prg1 + prg2 + bg + obj + rgb + bgcolor + objcolor + snd

def pngstring(a):
    w = 1024
    img = Image.new('L', (w, ceil(len(a) / w)))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/pac_and_pal.js
/*
 *	Pac & Pal
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = cat(z.read('pap1-3b.1d'), z.read('pap1-2b.1c'), z.read('pap3-1.1b'));
const prg2 = z.read('pap1-4.1k');
const bg = z.read('pap1-6.3c');
const obj = z.read('pap1-5.3f');
const rgb = z.read('pap1-6.4c');
const bgcolor = z.read('pap1-5.4e');
const objcolor = z.read('pap1-4.3l');
const snd = z.read('pap1-3.3m');

const rom = cat(prg1, prg2, bg, obj, rgb, bgcolor, objcolor, snd);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
